### ISSUE TITLE
feat: Implement LLM text detection stylometric features

### ISSUE THREAD

**Issue by craigtrim** [craigtrim/pystylometry#32]
## Overview

Based on research from ["Stylometry recognizes human and LLM-generated texts in short samples"](https://www.sciencedirect.com/science/article/pii/S0957417425026181) (also available on [arXiv](https://arxiv.org/pdf/2507.00838)), which demonstrates how stylometric features can distinguish between human-written and LLM-generated texts, several features should be implemented to support AI text detection use cases.

### Source Reference

> The paper demonstrates that traditional stylometric approaches can effectively detect LLM-generated text even in short samples (10 sentences). Key findings:
> - Tree-based models (Decision Trees, LightGBM) using stylometric features achieve high accuracy
> - Features include lexical, grammatical, syntactic, and punctuation patterns
> - StyloMetrix human-designed features combined with n-gram features perform well
> - Tests across multiple LLMs: GPT-3.5/4, LLaMA 2/3, Orca, Falcon
> - Paraphrasing generally increases (not decreases) detection rates

---

### Currently Implemented ✅

| Feature | Implementation |
|---------|----------------|
| POS tag ratios | `syntactic/pos_analysis.py` |
| Punctuation patterns | `character/character_metrics.py` |
| Sentence length statistics | `syntactic/sentence_stats.py` |
| Function words | `lexical/function_words.py` |
| N-gram entropy | `ngrams/entropy.py` |

### Not Implemented ❌

| Feature | Description |
|---------|-------------|
| Character n-gram features | Character-level n-gram frequency profiles |
| Word n-gram features | Word-level n-gram frequency profiles |
| POS n-gram features | Part-of-speech sequence patterns |
| Feature extraction for ML pipelines | Export features in ML-ready format |
| Text classifier interface | Wrapper for scikit-learn/LightGBM models |
| Human vs LLM detection utilities | Specialized detection pipeline |

---

### 1. Character N-gram Feature Extractor

**Description:** Extract character-level n-gram frequency profiles, which capture spelling patterns, morphological tendencies, and character-level regularities that differ between human and LLM text.

**Metrics to implement:**
- Character bigram frequencies (256+ features)
- Character trigram frequencies (top-k most frequent)
- Character 4-gram frequencies (top-k)
- Normalized frequency vectors
- TF-IDF weighted n-gram profiles

**Technical approach:**
```python
def character_ngram_features(
    text: str,
    n_range: Tuple[int, int] = (2, 4),
    top_k: int = 500,
    weighting: str = "frequency"  # or "tfidf"
) -> CharNgramResult:
    """
    Extract character n-gram feature vectors.
    
    Returns frequency profiles for stylometric analysis and ML pipelines.
    """
```

**Example output:**
```python
{
    "bigrams": {
        "th": 0.0312,
        "he": 0.0287,
        "in": 0.0245,
        "er": 0.0231,
        # ... top_k features
    },
    "trigrams": {
        "the": 0.0198,
        "ing": 0.0156,
        "and": 0.0134,
        # ...
    },
    "feature_vector": [0.0312, 0.0287, ...],  # ML-ready array
    "feature_names": ["bi_th", "bi_he", ...],
    "metadata": {
        "n_range": [2, 4],
        "top_k": 500,
        "total_ngrams_extracted": 1500
    }
}
```

---

### 2. Word N-gram Feature Extractor

**Description:** Extract word-level n-gram frequencies, capturing phrasal patterns and collocations that may differ between human and AI writing.

**Metrics to implement:**
- Word unigram frequencies (vocabulary profile)
- Word bigram frequencies (collocations)
- Word trigram frequencies (phrasal patterns)
- Skip-gram patterns (word pairs with gaps)
- Normalized and TF-IDF weighted vectors

**Technical approach:**
```python
def word_ngram_features(
    text: str,
    n_range: Tuple[int, int] = (1, 3),
    top_k: int = 1000,
    lowercase: bool = True,
    remove_stopwords: bool = False
) -> WordNgramResult:
    """
    Extract word n-gram feature vectors for stylometric analysis.
    """
```

**Example output:**
```python
{
    "unigrams": {
        "the": 0.071,
        "of": 0.035,
        "and": 0.028,
        # ...
    },
    "bigrams": {
        "of the": 0.0089,
        "in the": 0.0076,
        "to the": 0.0054,
        # ...
    },
    "trigrams": {
        "one of the": 0.0012,
        "as well as": 0.0009,
        # ...
    },
    "feature_vector": [...],
    "feature_names": [...],
    "vocabulary_size": 2847
}
```

---

### 3. POS N-gram Feature Extractor

**Description:** Extract part-of-speech tag sequence patterns, capturing syntactic regularities that may reveal authorship or generation source.

**Metrics to implement:**
- POS unigram distribution (tag frequencies)
- POS bigram sequences (tag transitions)
- POS trigram sequences
- Transition probability matrix
- Entropy of POS sequences

**Technical approach:**
```python
def pos_ngram_features(
    text: str,
    n_range: Tuple[int, int] = (1, 3),
    tagset: str = "universal"  # or "penn"
) -> POSNgramResult:
    """
    Extract POS tag n-gram features using spaCy.
    
    Captures syntactic patterns independent of vocabulary.
    """
```

**Example output:**
```python
{
    "unigrams": {
        "NOUN": 0.23,
        "VERB": 0.18,
        "DET": 0.12,
        "ADP": 0.11,
        # ...
    },
    "bigrams": {
        "DET_NOUN": 0.089,
        "NOUN_VERB": 0.067,
        "ADP_DET": 0.054,
        # ...
    },
    "trigrams": {
        "DET_NOUN_VERB": 0.034,
        "ADP_DET_NOUN": 0.028,
        # ...
    },
    "transition_matrix": [...],  # POS transition probabilities
    "sequence_entropy": 3.45,
    "metadata": {
        "tagset": "universal",
        "nlp_model": "en_core_web_sm"
    }
}
```

---

### 4. Stylometric Feature Aggregator for ML

**Description:** Aggregate multiple stylometric features into a single feature vector suitable for machine learning pipelines, following the StyloMetrix approach.

**Metrics to implement:**
- All lexical metrics (TTR, MTLD, Yule, etc.)
- All readability metrics (Flesch, SMOG, etc.)
- All syntactic metrics (POS ratios, sentence stats)
- All character metrics
- N-gram features (character, word, POS)
- Configurable feature selection

**Technical approach:**
```python
def extract_stylometric_features(
    text: str,
    feature_sets: List[str] = ["all"],
    return_format: str = "dict"  # or "array", "dataframe"
) -> Union[Dict, np.ndarray, pd.DataFrame]:
    """
    Extract comprehensive stylometric feature vector.
    
    Combines all pystylometry metrics into ML-ready format.
    Feature sets: "lexical", "readability", "syntactic", 
                  "character", "ngram", "all"
    """
```

**Example output:**
```python
{
    "features": {
        # Lexical
        "ttr": 0.45,
        "mtld": 78.3,
        "yule_k": 124.5,
        # Readability
        "flesch_reading_ease": 62.4,
        "smog_index": 10.2,
        # Syntactic
        "noun_ratio": 0.23,
        "verb_ratio": 0.18,
        "mean_sentence_length": 18.5,
        # Character
        "avg_word_length": 4.73,
        "punctuation_density": 0.12,
        # N-grams (selected)
        "char_bigram_th": 0.031,
        "word_bigram_of_the": 0.008,
        # ...
    },
    "feature_vector": np.array([0.45, 78.3, 124.5, ...]),
    "feature_names": ["ttr", "mtld", "yule_k", ...],
    "n_features": 150,
    "metadata": {
        "feature_sets": ["lexical", "readability", "syntactic", "character"],
        "text_length": 5000
    }
}
```

---

### 5. Human vs LLM Detection Pipeline

**Description:** Implement a specialized pipeline for detecting LLM-generated text using stylometric features, based on the research findings.

**Metrics to implement:**
- Pre-trained feature importance weights (from research)
- SHAP-based feature explanation
- Confidence scores with uncertainty estimates
- Per-feature contribution to prediction
- Support for multiple LLM detection (GPT vs LLaMA vs human)

**Technical approach:**
```python
class LLMDetector:
    """
    Stylometry-based LLM text detector.
    
    Uses tree-based models with stylometric features
    to distinguish human from LLM-generated text.
    """
    
    def __init__(self, model_type: str = "lightgbm"):
        pass
    
    def extract_features(self, text: str) -> np.ndarray:
        """Extract stylometric features for classification."""
        pass
    
    def predict(self, text: str) -> LLMDetectionResult:
        """Predict whether text is human or LLM-generated."""
        pass
    
    def explain(self, text: str) -> Dict:
        """SHAP-based explanation of prediction."""
        pass
```

**Example output:**
```python
{
    "prediction": "llm_generated",
    "confidence": 0.87,
    "probabilities": {
        "human": 0.13,
        "gpt4": 0.52,
        "gpt35": 0.18,
        "llama": 0.12,
        "other_llm": 0.05
    },
    "top_discriminating_features": [
        {"feature": "sentence_length_std", "contribution": 0.15, "value": 2.3},
        {"feature": "comma_density", "contribution": 0.12, "value": 0.08},
        {"feature": "pos_bigram_DET_NOUN", "contribution": 0.09, "value": 0.091},
        # ...
    ],
    "explanation": "Text shows unusually consistent sentence lengths and punctuation patterns typical of LLM output.",
    "metadata": {
        "model": "lightgbm",
        "n_features_used": 150,
        "min_text_length_required": 100
    }
}
```

---

### File Structure

```
pystylometry/
├── ngrams/
│   ├── character_ngrams.py    # Feature 1
│   ├── word_ngrams.py         # Feature 2
│   └── pos_ngrams.py          # Feature 3
├── ml/
│   ├── __init__.py
│   ├── feature_extractor.py   # Feature 4
│   ├── llm_detector.py        # Feature 5
│   └── models/
│       └── pretrained/        # Pre-trained model weights
```

### Dependencies

- scikit-learn for ML utilities
- LightGBM for gradient boosting (optional)
- SHAP for model explanations (optional)
- spaCy for POS tagging

### Priority Order

1. **High Priority** (Core features for LLM detection):
   - Character N-gram Feature Extractor
   - POS N-gram Feature Extractor
   - Stylometric Feature Aggregator

2. **Medium Priority**:
   - Word N-gram Feature Extractor

3. **Lower Priority** (Requires ML infrastructure):
   - Human vs LLM Detection Pipeline

---

## References

- [Stylometry recognizes human and LLM-generated texts in short samples](https://www.sciencedirect.com/science/article/pii/S0957417425026181) - ScienceDirect
- [arXiv preprint](https://arxiv.org/pdf/2507.00838)
- [StyloMetrix](https://github.com/ZILiAT-NASK/StyloMetrix) - Polish stylometric feature extraction tool referenced in the paper

**Comment by craigtrim** [comment]
## Implementation Audit — January 28, 2026

Full codebase review against the proposed features in this issue. Results below.

---

### Issue's "Currently Implemented" List — Verified ✅

All five items the issue listed as already implemented are confirmed present and functional:

| Feature | File | Notes |
|---------|------|-------|
| POS tag ratios | `syntactic/pos_ratios.py` | Noun, verb, adj, adv ratios + lexical density + function word ratio |
| Punctuation patterns | `character/character_metrics.py` | Density, variety, 30+ punctuation types |
| Sentence length statistics | `syntactic/sentence_stats.py` | Mean, std, range, min, max, count with Distribution objects |
| Function words | `lexical/function_words.py` | 6 categories: determiners (78), prepositions (100+), conjunctions (40+), pronouns, aux verbs, particles |
| N-gram entropy | `ngrams/entropy.py` | Shannon entropy + perplexity for word and character n-grams |

---

### Issue's "Not Implemented" List — Reassessed

**3 of 6 are now implemented.** The issue's gap analysis is outdated:

| Feature | Issue Status | Actual Status | Details |
|---------|-------------|---------------|---------|
| Character n-gram features | ❌ | ✅ **Implemented** | `ngrams/extended_ngrams.py` — char trigrams, 4-grams, entropy, top-100 frequency distributions |
| Word n-gram features | ❌ | ✅ **Implemented** | `ngrams/extended_ngrams.py` — word trigrams, 4-grams, 2-skipgrams, 3-skipgrams, entropy |
| POS n-gram features | ❌ | ✅ **Implemented** | `ngrams/extended_ngrams.py` — POS trigrams, 4-grams, entropy via spaCy |
| Feature extraction for ML pipelines | ❌ | ⚠️ **Partial** | 44 compute functions with typed results exist, but no unified aggregator that outputs a single ML-ready feature vector |
| Text classifier interface | ❌ | ❌ **Not implemented** | No scikit-learn/LightGBM wrapper |
| Human vs LLM detection utilities | ❌ | ⚠️ **Partial** | `consistency/drift.py` detects `suspiciously_uniform` patterns (a signal of AI generation), but there is no unified LLM detection pipeline |

---

### What Could Be Augmented

#### 1. N-gram feature gaps (character bigrams, word unigrams/bigrams)

`extended_ngrams.py` covers trigrams and 4-grams but **does not include**:
- **Character bigrams** — the most discriminative character n-gram size per the paper
- **Word unigrams** (vocabulary frequency profile)
- **Word bigrams** (collocations)
- **POS bigrams** (tag transition pairs)

`entropy.py` computes bigram entropy but does not return the frequency distributions themselves. The paper specifically relies on bigram-level features for LLM detection.

**Recommendation:** Extend `extended_ngrams.py` to cover n=2 for all three types (char, word, POS), or add a lower-order counterpart.

#### 2. TF-IDF weighting for n-gram profiles

All n-gram features use raw frequency counts. The issue proposes TF-IDF weighted n-gram profiles, which would require a reference corpus. This is not currently supported.

#### 3. Stylometric Feature Aggregator (Proposed Feature #4)

This is the highest-value missing piece. All 44 compute functions return independent typed results. There is no single function that:
- Runs a configurable set of analyses
- Flattens results into a single `Dict[str, float]` or `np.ndarray`
- Returns `feature_names` for downstream ML use

This would bridge pystylometry's feature extraction to any ML framework without coupling to one.

#### 4. LLM Detection Pipeline (Proposed Feature #5)

The building blocks exist (`drift.py`, all feature extractors), but there is no:
- Unified pipeline that runs features → model → prediction
- Pre-trained model weights
- SHAP integration for explainability

This is architecturally the largest addition and introduces ML framework dependencies (scikit-learn, LightGBM, SHAP). It may warrant being a separate package or optional extra.

---

### Summary

| Category | Count |
|----------|-------|
| Features listed as not implemented that **are now implemented** | 3 |
| Features that need **augmentation** (bigrams, TF-IDF) | 2 |
| Features that are **genuinely missing** | 2 (Feature Aggregator + LLM Detection Pipeline) |

The issue's gap analysis should be updated. The core n-gram extractors (char, word, POS) are all present. The remaining work is:

1. **Low effort:** Add bigram-level n-gram extraction alongside existing trigram/4-gram support
2. **Medium effort:** Build a feature aggregator that flattens all 44 compute functions into a single ML-ready vector
3. **High effort:** Build or scope out the LLM detection pipeline (consider whether this belongs in pystylometry or a separate package)
